src/workerd/api/pyodide/pyodide.h

```diff
@@ -78,12 +79,78 @@ class PyodideMetadataReader : public jsg::Object {
   }
 };
 
+// A loaded bundle of artifacts for a particular script id. It can also contain V8 version and
+// CPU architecture-specific artifacts. The logic for loading these is in getArtifacts.
+class ArtifactBundler : public jsg::Object {
+public:
+  ArtifactBundler(kj::Maybe<kj::Array<kj::byte>> existingSnapshot,
+      kj::Function<kj::Promise<bool>(kj::Array<kj::byte> snapshot)> uploadMemorySnapshotCb)
+      : existingSnapshot(kj::mv(existingSnapshot)),
+        uploadMemorySnapshotCb(kj::mv(uploadMemorySnapshotCb)),
+        hasUploaded(false) {};
+
+  ArtifactBundler()
+      : existingSnapshot(kj::none),
+        uploadMemorySnapshotCb(kj::none),
+        hasUploaded(false) {};
+
+  jsg::Promise<bool> uploadMemorySnapshot(jsg::Lock& js, kj::Array<kj::byte> snapshot) {
+    // Prevent multiple uploads.
+    if (hasUploaded) {
+      return js.rejectedPromise<bool>(
+          js.typeError("This ArtifactBundle has already uploaded a memory snapshot"));
+    }
+
+    // TODO(later): Only upload if `snapshot` isn't identical to `existingSnapshot`.
+
+    auto& cb = JSG_REQUIRE_NONNULL(uploadMemorySnapshotCb, TypeError, "ArtifactBundler is disabled");
```

Comment:
This should probably return a rejected promise also, instead of throwing synchronously